clang/runtime/dpct-rt/include/dpct/codepin/codepin.hpp

```diff
@@ -1,83 +1,212 @@
+
+inline std::map<void *, uint32_t> &get_ptr_size_map() {
+  static std::map<void *, uint32_t> ptr_size_map;
+  return ptr_size_map;
+}
+
+inline uint32_t get_ptr_size_in_bytes_from_map(void *ptr) {
+  const std::map<void *, uint32_t> &ptr_size_map = get_ptr_size_map();
+  const auto &it = ptr_size_map.find(ptr);
+  return (it != ptr_size_map.end()) ? it->second : 0;
+}
+
+inline bool is_dev_ptr(void *p) {
 #ifdef __NVCC__
-#include <cuda_runtime.h>
+  cudaPointerAttributes attr;
+  cudaPointerGetAttributes(&attr, p);
+  if (attr.type == cudaMemoryTypeDevice)
+    return true;
+  return false;
 #else
-#include <dpct/dpct.hpp>
-#include <sycl/sycl.hpp>
+  dpct::pointer_attributes attributes;
+  attributes.init(p);
+  if (attributes.get_device_pointer() != nullptr)
+    return true;
+  return false;
 #endif
-namespace dpct {
-namespace experimental {
+}
 
+template <class T>
+class TT<T, typename std::enable_if<std::is_pointer<T>::value>::type> {
+public:
+  static void dump(std::ostream &ss, T value,
+                   dpct::experimental::StreamType stream) {
+    if (ptr_unique.find(value) != ptr_unique.end()) {
+      return;
+    }
+    ptr_unique.insert(value);
+    ss << "{\"Type\":\"Pointer\",\"Data\":[";
+    int size = get_ptr_size_in_bytes_from_map(value);
+    size = size == 0 ? 1 : size / sizeof(*value);
+    using PointedType =
+        std::remove_reference_t<std::remove_cv_t<std::remove_pointer_t<T>>>;
+    if (is_dev_ptr(value)) {
+      PointedType *h_data = new PointedType[size];
 #ifdef __NVCC__
-inline void synchronize(cudaStream_t stream) { cudaStreamSynchronize(stream); }
+      cudaMemcpyAsync(h_data, value, size * sizeof(PointedType),
+                      cudaMemcpyDeviceToHost, stream);
+      cudaStreamSynchronize(stream);
+#else
+      stream->memcpy(h_data, value, size * sizeof(PointedType)).wait();
+#endif
+      for (int i = 0; i < size; ++i) {
+        dpct::experimental::detail::TT<PointedType>::dump(ss, *(h_data + i),
+                                                          stream);
+        if (i != size - 1)
+          ss << ",";
+      }
+      delete[] h_data;
+    } else {
+      for (int i = 0; i < size; ++i) {
+        dpct::experimental::detail::TT<PointedType>::dump(ss, *(value + i),
+                                                          stream);
+        if (i != size - 1)
+          ss << ",";
+      }
+    }
 
-/// Generate API check point prolog.
-/// \param api_name The UID of the function call.
-/// \param stream The CUDA stream to synchronize the command execution.
-/// \param args The schema string and variable value pair list.
-template <class... Args>
-void gen_prolog_API_CP(const std::string &api_name, cudaStream_t stream,
-                       Args... args) {
-  synchronize(stream);
-  dpct::experimental::detail::gen_log_API_CP(api_name, args...);
+    ss << "]}";
+  }
+};
+
+template <class T>
+class TT<T, typename std::enable_if<std::is_array<T>::value>::type> {
```

Comment:
How about change TT to DataSer / (Data serialize for any type of Data),  and add comment for the util class.